File: Labs/Lab 6/Lab 6.py
def canFinish(self, numCourses, prerequisites):
        graph = [[] for _ in range(numCourses)]
        visited = [0] * numCourses

        for prerequisite in prerequisites:
            graph[prerequisite[0]].append(prerequisite[1])
        
        def hasCycle(course):
            if visited[course] == 1:
                return True
            if visited[course] == 2:
                return False
            
            visited[course] = 1
            for neighbor in graph[course]:
                if hasCycle(neighbor):
                    return True
            
            visited[course] = 2
            return False
        
        for course in range(numCourses):
            if hasCycle(course):
                return False
        return True
        

File: Labs/Lab 6/Lab 6.java
public boolean canFinish(int numCourses, int[][] prerequisites) {
       List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] prerequisite : prerequisites) {
            graph.get(prerequisite[0]).add(prerequisite[1]);
        }

        int[] visited = new int[numCourses]; 

        for (int i = 0; i < numCourses; i++) {
            if (hasCycle(graph, visited, i)) {
                return false;
            }
        }

        return true;    
}

private boolean hasCycle(List<List<Integer>> graph, int[] visited, int course) {
        if (visited[course] == 1) { 
            return true;
        }
        if (visited[course] == 2) { 
            return false;
        }

        visited[course] = 1; 
        for (int neighbor : graph.get(course)) {
            if (hasCycle(graph, visited, neighbor)) {
                return true;
            }
        }
        visited[course] = 2; 

        return false;
}
